Support for CentOS 7

ebeigarts opened the pull request:
This pull request adds support for CentOS 7 and also builds RPM packages. Fixes #478.

Currently `rpm.mk` looks very similar to `deb.mk`, maybe this could be refactored in the future to reuse code more effectively.

Installation and testing using DigitalOcean with Centos 7.2:

```
package_cloud push ebeigarts/dokku/el/7 *.rpm

wget https://raw.githubusercontent.com/ebeigarts/dokku/centos/bootstrap.sh
sed -i "s/dokku\/dokku/ebeigarts\/dokku/" bootstrap.sh
DOKKU_TAG=v0.7.2 bash bootstrap.sh

ssh-keygen -f ~/.ssh/id_rsa -t rsa -N ''

dokku apps:create node-js-sample
dokku ssh-keys:add node-js-sample ~/.ssh/id_rsa.pub
dokku domains:add node-js-sample localhost

git clone https://github.com/heroku/node-js-sample
cd node-js-sample/
git remote add dokku dokku@localhost:node-js-sample
git push dokku master

curl http://localhost/ # => Hello World!
```

josegonzalez commented:
I'll discuss this changeset with @michaelshobbs this week, but good work so far! The only thing I can think of is that the `bootstrap.sh` should be modified to handle Centos7 (and no other versions of course).

josegonzalez commented:
Without reviewing the code, a few things:

- None of the core maintainers use CentOS - nor do we have any desire to.
- We'll bring you - @ebeigarts - up in any issues referencing CentOS support.
- These packages will not be pushed to any official CentOS/RHEL repository, and any such packages will be considered unofficial and unsupported. All packages will be available in our packagecloud repository, as we do for debian packages.
- This integration would be considered experimental and not supported by the core team.
- Should this integration become a hassle to upkeep, we'll drop it from the core.
- Needs a few more touch-ups to the release process documentation.

@ebeigarts Assuming this integration goes well, we'll have you as point-person for the package, and likely eventually commit bit.

josegonzalez commented on plugins/nginx-vhosts/install:
```diff
@@ -17,6 +17,11 @@ case "$DOKKU_DISTRO" in
+
+  centos)
+    echo "%dokku ALL=(ALL) NOPASSWD:/usr/bin/systemctl reload nginx, /usr/sbin/nginx -t" > /etc/sudoers.d/dokku-nginx
+    echo "Defaults:dokku !requiretty" >> /etc/sudoers.d/dokku-nginx
```
Why this?

ebeigarts replied:
CentOS has `requiretty` by default in `/etc/sudoers` and requires that `sudo` [command has tty](https://www.shell-tips.com/2014/09/08/sudo-sorry-you-must-have-a-tty-to-run-sudo/). I was getting `sudo: sorry, you must have a tty to run sudo` error when using `git push`, dokku couldn't restart the nginx with sudo.

josegonzalez commented on rpm.mk:
```diff
@@ -0,0 +1,171 @@
+	cp -rf /tmp/tmp/herokuish /tmp/build/var/lib/herokuish
+
+	@echo "-> Creating $(HEROKUISH_RPM_PACKAGE_NAME)"
+	sudo fpm -t rpm -s dir -C /tmp/build -n herokuish \
```
Can fpm for rpms be run under ubuntu?

ebeigarts replied:
Yes, it just requires `apt-get install rpm` which I added in `rpm-setup`.

josegonzalez commented on rpm.mk:
```diff
@@ -0,0 +1,171 @@
+		-a $(RPM_ARCHITECTURE) \
+		-p $(HEROKUISH_RPM_PACKAGE_NAME) \
+		--depends 'docker-engine >= 1.9.1' \
+		--depends 'sudo' \
```
This is a debianism, but worth keeping in case CentOs screws the pooch or there is a user out there who thinks they are slick.

josegonzalez commented on rpm.mk:
```diff
@@ -0,0 +1,171 @@
+		--depends 'gliderlabs-sigil' \
+		--depends 'docker-engine >= 1.9.1' \
+		--depends 'bind-utils' \
+		--depends 'nginx' \
```
What version of nginx comes with CentOs 7?

ebeigarts replied:
none. [EPEL](https://fedoraproject.org/wiki/EPEL) repository includes nginx 1.10.1. That is why I had to `yum install -y epel-release`

josegonzalez commented on rpm/dokku.postinst:
```diff
@@ -0,0 +1,60 @@
+  fi
+}
+
+echo "Setting up dokku user"
```
Any reason why this section isn't all in a function at least? The debian one isn't because of the diff, though this all should be.

ebeigarts replied:
This is mostly copy/paste from debian/postinst.

josegonzalez commented on rpm/herokuish.postinst:
```diff
@@ -0,0 +1,16 @@
+
+sleep 5
+
+count=`sudo docker images | grep gliderlabs/herokuish | wc -l`
```
Can we use `$()` instead of backticks everywhere?

ebeigarts replied:
This was taken from https://github.com/dokku/dokku/blob/master/deb.mk#L87-L95. Do we need to change that also in deb packages?

josegonzalez replied:
This file isnt a Makefile.

josegonzalez commented on rpm/dokku.postinst:
```diff
@@ -0,0 +1,60 @@
+dokku plugin:install --core
+
+rm -f ${DOKKU_ROOT}/VERSION
+cp ${DOKKU_LIB_ROOT}/STABLE_VERSION ${DOKKU_ROOT}/VERSION
```
Please quote all variables and make these files pass shellcheck.

ebeigarts replied:
Do we need to quote this also in `debian/postinst`?

josegonzalez replied:
I'm pretty sure it is, if not, we should, but separate pr please.

michaelshobbs commented:
This should probably be in the contributing doc if it isn't already but we try to follow this style guide as best we can. 

https://github.com/progrium/bashstyle

ebeigarts commented:
Everythin else sounds good, I will change `bootstrap.sh` to support centos7

ebeigarts commented:
I have added support for CentOS in `bootstrap.sh` and also improved code using shellcheck suggestions. I also switched to using `/etc/os-release` that is available on Ubuntu/Debian/Arch/CentOS and others by default (`yum install redhat-lsb-core` is not required this way)

ebeigarts commented on bootstrap.sh:
```diff
@@ -33,10 +29,15 @@ ensure-environment() {
+    debian|ubuntu)
+      apt-get update -qq > /dev/null
+      if [[ "$DOKKU_DISTRO_VERSION" == "12.04" ]]; then
+        apt-get -qq -y install python-software-properties
```
Is this still needed? I saw that 12.04 is not supported anymore.

josegonzalez replied:
Yes, I would leave this here for now.

ebeigarts commented:
@josegonzalez what do you mean by rpm docs page? Something like http://dokku.viewdocs.io/dokku/getting-started/install/debian/ or release instructions? Release instructions are already documented - rpms are built using `vagrant up build`.

josegonzalez commented:
Yes, like that page.

ebeigarts commented:
@josegonzalez I have added [rpm docs](https://github.com/ebeigarts/dokku/blob/centos/docs/getting-started/install/rpm.md). anything else?

josegonzalez commented on docs/getting-started/install/rpm.md:
```diff
@@ -0,0 +1,16 @@
+# RPM Package Installation Notes
+
```
Can you add a blockquote like so:

```
> New as of 0.8
```

ebeigarts replied:
done

ebeigarts commented:
@josegonzalez is there anything holding this back from merging?

josegonzalez commented:
@ebeigarts it'll go in the next minor.
